satellite/gc/service.go

```diff
@@ -0,0 +1,136 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package gc
+
+import (
+	"context"
+	"time"
+
+	"github.com/zeebo/errs"
+	"go.uber.org/zap"
+
+	"storj.io/storj/pkg/auth/signing"
+	"storj.io/storj/pkg/overlay"
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/storj"
+	"storj.io/storj/pkg/transport"
+	"storj.io/storj/uplink/piecestore"
+)
+
+var (
+	// Error defines the gc service errors class
+	Error = errs.Class("gc service error")
+)
+
+// Config contains configurable values for garbage collection
+type Config struct {
+	Interval time.Duration `help:"how frequently garbage collection filters should be sent to storage nodes" releaseDefault:"168h" devDefault:"10m"`
+	Active   bool          `help:"set if garbage collection is actively running or not" releaseDefault:"true" devDefault:"true"`
+	// value for InitialPieces currently based on average pieces per node
+	InitialPieces     int64   `help:"the initial number of pieces expected for a storage node to have, used for creating a filter" releaseDefault:"400000" devDefault:"10"`
+	FalsePositiveRate float64 `help:"the false positive rate used for creating a filter" releaseDefault:"0.1" devDefault:"0.1"`
+}
+
+// Service implements the garbage collection service
+type Service struct {
+	log             *zap.Logger
+	config          Config
+	transport       transport.Client
+	overlay         overlay.DB
+	lastSendTime    time.Time
+	lastPieceCounts map[storj.NodeID]int
+}
+
+// NewService creates a new instance of the gc service
+func NewService(log *zap.Logger, config Config, transport transport.Client, overlay overlay.DB) *Service {
+	return &Service{
+		log:             log,
+		config:          config,
+		transport:       transport,
+		overlay:         overlay,
+		lastPieceCounts: make(map[storj.NodeID]int),
+	}
+}
+
+// NewPieceTracker instantiates a piece tracker
+func (service *Service) NewPieceTracker() *PieceTracker {
+	// Creation date of the gc bloom filter - the storage nodes shouldn't delete any piece newer than this.
+	filterCreationDate := time.Now().UTC()
+
+	if !service.config.Active || filterCreationDate.Before(service.lastSendTime.Add(service.config.Interval)) {
+		return nil
+	}
+
+	return &PieceTracker{
+		log:                service.log.Named("piecetracker"),
+		filterCreationDate: filterCreationDate,
+		initialPieces:      service.config.InitialPieces,
+		falsePositiveRate:  service.config.FalsePositiveRate,
+		retainInfos:        make(map[storj.NodeID]*RetainInfo),
+		pieceCounts:        service.lastPieceCounts,
+		overlay:            service.overlay,
+	}
+}
+
+// Send sends the piece retain requests to all storage nodes
+func (service *Service) Send(ctx context.Context, pieceTracker *PieceTracker, cb func()) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	service.lastSendTime = time.Now().UTC()
+
+	go func() {
```

Comment:
Don't make unbounded number goroutines.

Reply:
Would it be preferable to block the next iteration of checker until all the bloom filters are sent? Or is there a better way to send the filters in an async way?

Reply:
By "unbounded number goroutines", do you just mean that we could have an arbitrarily large number of retain requests in our map?

Reply:
We will have thousands of storage nodes. If we start sending bloom filter to all of them in parallel, this may hog the bandwidth capacity of the satellite and disrupt the operation of other services.

We should keep the GC process low profile. Perhaps, we should upload bloom filter just to a (configurable) handful of storage nodes in parallel.

Reply:
it should actually be sending the bloom filters one by one, not all in parallel. The goroutine just splits the send loop off from the checker process.

Reply:
We shouldn't block the next iteration of checker since the sending could take significant amount of time. 

The unbounded means that you shouldn't call `go` unless it's obvious how they are waited for. Currently when you call `Send` there's no way to wait for their completion internally for the service in close.

Also, if it's sending one by one, then it could end up taking a lot of time.

Reply:
In your view, would it be fine to use a `sync2` `Limiter` and add a config value for number of concurrent sends?

Reply:
There shouldn't be any concurrent sends.

Reply:
If there aren't concurrent sends, then we will have to be okay with it taking a lot of time because we are sending one by one.

Reply:
Except in a testing environment, I do not think we will need to worry about creating/sending filters multiple times in the same window, as garbage collection is configured for once a week right now

Reply:
I mean there shouldn't be two concurrent sending of piecetracker info. The retain requests should still happen in parallel.

Reply:
We shouldn't use timing for correctness, this causes subtle bugs. E.g. someone in the future tries the configuration with 1h.

Reply:
Ok so what about the following implementation?
1. Create `sendLimiter`, which can be configured to do some number of concurrent retain request sends
2. Have some cancellable context attached to the garbage collection service. For every retain request we send, we pass this context into `sendLimiter.Go` along with the function for sending the retain request.
3. When we attempt to send all retain requests, we first cancel the context passed into the limiter (just in case we are currently sending any retain requests from a previous call). We update the service with a new cancellable context before adding extra goroutines to the limiter

Reply:
The logic for looping should be pulled out of checker, there should be registration, proper observers, joining into looping etc. To clean this up.